src/commons/eventsourcing/wal.rs

```diff
@@ -0,0 +1,352 @@
+    /// internal data of the aggregate.
+    ///
+    /// Note the event is moved. This is done because we want to avoid
+    /// doing additional allocations where we can.
+    fn apply(&mut self, set: WalSet<Self>);
+
+    /// Processes a command. I.e. validate the command, and return a list of
+    /// events that will result in the desired new state, but do not apply
+    /// these events here.
+    ///
+    /// The command is moved, because we want to enable moving its data
+    /// without reallocating.
+    fn process_command(&self, command: Self::Command) -> Result<Vec<Self::Change>, Self::Error>;
+}
+
+//------------ WalCommand ----------------------------------------------------
+
+pub trait WalCommand: fmt::Display {
+    fn handle(&self) -> &MyHandle;
+}
+
+//------------ WalEvent ------------------------------------------------------
+
+pub trait WalChange: fmt::Display + Eq + PartialEq + Send + Sync + Storable {}
+
+//------------ WalSet --------------------------------------------------------
+
+/// Describes a set of "write-ahead" changes affecting the specified revision.
+/// Meaning that it can only be applied if the type is of the given revision, and
+/// it will get this revision + 1 after it has been applied.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct WalSet<T: WalSupport> {
+    revision: u64,
+    summary: String,
+    changes: Vec<T::Change>,
+}
+
+impl<T: WalSupport> WalSet<T> {
+    pub fn into_changes(self) -> Vec<T::Change> {
+        self.changes
+    }
+}
+
+//------------ WalStore ------------------------------------------------------
+
+/// This type is responsible for loading / saving and updating [`WalSupport`]
+/// capable types.
+///
+/// This is similar to how [`AggregateStore`] is used to manage [`Aggregate`]
+/// types. However, there are some important differences:
+/// - Commands and events for a change are saved as a single file.
+/// - Old commands and events are no longer relevant and will be removed.
+///   (we may want to support archiving those in future).
+/// - We do not have any listeners in this case.
+/// - We cannot replay [`WriteAheadSupport`] types from just events, we
+///   *always* need to start with an existing snapshot.
+#[derive(Debug)]
+pub struct WalStore<T: WalSupport> {
+    kv: KeyValueStore,
+    cache: RwLock<HashMap<MyHandle, Arc<T>>>,
+}
+
+impl<T: WalSupport> WalStore<T> {
+    /// Creates a new store using a disk based keystore for the given data
+    /// directory and namespace (directory).
+    pub fn disk(krill_data_dir: &Path, name_space: &str) -> WalStoreResult<Self> {
+        let mut path = krill_data_dir.to_path_buf();
+        path.push(name_space);
+
+        let kv = KeyValueStore::disk(krill_data_dir, name_space)?;
+        let cache = RwLock::new(HashMap::new());
+
+        Ok(WalStore { kv, cache })
+    }
+
+    /// Warms up the store: caches all instances.
+    pub fn warm(&self) -> WalStoreResult<()> {
```

Comment:
Minor: This is doing something (i.e. warming up) rather than being "warm" so wouldn't it be better named `warmup()` or similar?

Reply:
I prefer to keep it like this. This is consistent with `AggregateStore::warm`. Fwiw.. "warm" was intended as a verb, and the imperative at that. For being warm I would use something like "is_warm" or "warmed" (i.e. past tense, not the imperative).

Reply:
To be clear - I think `warmup` would have been better, but I don't think that `warm` on its own is too problematic. I am happy to change after all the stacked PRs are merged - because I fear conflicts.